core/frontend/src/render/FrameStats.ts

```diff
@@ -0,0 +1,133 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Rendering
+ */
+
+import { BeEvent } from "@bentley/bentleyjs-core";
+
+/** Describes timing statistics for a single rendered frame. Aside from `frameId`, `totalFrameTime`, and `sceneTime`, the other entries may represent operations that are not performed every frame and may contain an expected value of zero.
+ * @note By default, the display system does not render frames continuously. The display system will render a new frame only when the view changes. Therefore, the data contained within this interface cannot directly be used to compute a representative framerate.
+ * @alpha
+ */
+export interface FrameStats {
+  /** A unique number identifying the frame to which these statistics belong. */
+  readonly frameId: number;
+  /** The CPU time in milliseconds spent rendering the frame. */
+  readonly totalFrameTime: number;
+  /** The CPU time in milliseconds spent setting up the scene. The includes the following:
+   * - performing animations
+   * - setting up from view
+   * - setting the hilite set
+   * - overriding feature symbology
+   * - creating and changing the scene if invalid
+   * - validating render plan
+   * - adding decorations
+   * - processing flash
+   */
+  readonly sceneTime: number;
+  /** The CPU time in milliseconds spent rendering opaque geometry. */
+  readonly opaqueTime: number;
+  /** The CPU time in milliseconds spent rendering translucent geometry. */
+  readonly translucentTime: number;
+  /** The CPU time in milliseconds spent rendering overlays. */
+  readonly overlaysTime: number;
+  /** The CPU time in milliseconds spent rendering the solar shadow map. */
+  readonly shadowsTime: number;
+  /** The CPU time in milliseconds spent rendering both planar and volume classifiers. */
+  readonly classifiersTime: number;
+  /** The CPU time in milliseconds spent applying screenspace effects. */
+  readonly screenspaceEffectsTime: number;
+  /** The CPU time in milliseconds spent rendering background geometry including backgrounds, skyboxes, and background maps. */
+  readonly backgroundTime: number;
+}
+
+/** An event which will be raised when a new frame statistics object is available. The listeners will receive that frame statistics object.
+ * @see [[Viewport.enableFrameStatsListener]]
+ * @alpha
+ */
+export type OnFrameStatsReadyEvent = BeEvent<(frameStats: FrameStats) => void>;
+
+/** @internal */
+export class FrameStatsCollector {
+  private _onFrameStatsReady?: OnFrameStatsReadyEvent;
+  private _frameStatsMap = new Map<string, number>();

```

Comment:
What motivated the change to a map?

Reply:
FrameStats is now immutable so it cannot directly accumulate timings.  Is there a better way you can think of?  Open to suggestions.

Reply:
You can either define FrameStats with mutable member and pass a `Readonly<FrameStats>` to your callback;
or you can define `type MutableFrameStats = { -readonly [K in keyof FrameStats]: FrameStats[K] }` and give FrameStatsCollector a property of that type.

Reply:
I changed this to the follow pattern of a FrameStats with mutable members and passing a `Readonly<FrameStats>` to the event listeners.